File: lib/py/uapi/internal/validation/ValidateHeaders.py
from typing import List, Dict
from uapi.internal.types.UFieldDeclaration import UFieldDeclaration
from uapi.internal.types.UFn import UFn
from uapi.internal.validation.ValidationFailure import ValidationFailure


def validate_headers(headers: Dict[str, object], parsed_request_headers: Dict[str, 'UFieldDeclaration'], function_type: 'UFn') -> List['ValidationFailure']:
    validation_failures = []

    for header, header_value in headers.items():
        field = parsed_request_headers.get(header)
        if field:
            this_validation_failures = field.type_declaration.validate(
                header_value, None, function_type.name, [])
            this_validation_failures_path = [
                ValidationFailure([header] + e.path, e.reason, e.data)
                for e in this_validation_failures
            ]
            validation_failures.extend(this_validation_failures_path)

    return validation_failures

File: lib/java/src/main/java/uapi/internal/validation/ValidateHeaders.java
package uapi.internal.validation;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UFn;

public class ValidateHeaders {
    public static List<ValidationFailure> validateHeaders(
            Map<String, Object> headers, Map<String, UFieldDeclaration> parsedRequestHeaders, UFn functionType) {
        final var validationFailures = new ArrayList<ValidationFailure>();

        for (final var entry : headers.entrySet()) {
            final var header = entry.getKey();
            final var headerValue = entry.getValue();
            final var field = parsedRequestHeaders.get(header);
            if (field != null) {
                final var thisValidationFailures = field.typeDeclaration.validate(headerValue, null, functionType.name);
                final var thisValidationFailuresPath = thisValidationFailures.stream()
                        .map(e -> {
                            final var path = new ArrayList<>(e.path);
                            path.add(0, header);

                            return new ValidationFailure(path, e.reason, e.data);
                        }).toList();
                validationFailures.addAll(thisValidationFailuresPath);
            }
        }

        return validationFailures;
    }
}
